run_generator2 rust call and compact conditions data structure

arvidn opened the pull request:
This patch makes `chia-blockchain` use the new interface to the rust implementation of `run_generator()`.

The main improvement is to use a fixed struct containing all conditions, rather than a list of "arbitrary" conditions (where duplicates and redundant conditions are allowed).

This data structure simplifies the handling and checking of the conditions, and making it more straight-forward. It also reflects the internal data structure in the rust implementation.

It removes boiler plate code and opens the door to further performance improvements (by using the rust type as a streamable type for instance).

For example, this simpler condition summary object has a single field for the sum of all `RESERVE_FEE` conditions. Similarly for `ASSERT_HEIGHT_ABSOLUTE` also only requires a single field (the most strict one). Making this data structure compact on the rust side improves performance (by copying and allocating less memory) and simplifies checking the conditions, since we can have a single check, rather than a linear scan over all conditions.

arvidn committed:
use run_generator2 rust call and compact spend bundle conditions data structure pervasively.

mariano54 reviewed (commented):
It doens't seem like a huge change, it's mostly simplifications. I haven't reviewed it in detail though

arvidn commented:
I intend to get back to this once higher priority tasks complete. Perhaps I will split it up into two stages to make it easier to land

richardkiss commented on chia/types/spend_bundle_conditions.py:
```diff
@@ -0,0 +1,29 @@
+    height_relative: Optional[uint32]
+    seconds_relative: uint64
+    create_coin: List[Tuple[bytes32, uint64, bytes]]
+    agg_sig_me: List[Tuple[bytes48, bytes]]
```
I agree with Arvid that deferring is reasonable. I've thought a lot about the issues of parsing and rust and python data structures, and I believe in a lot of cases we want to do "dumb" deserialization since there may be instances where we don't want to validate anything (such as getting blobs off disk, which have presumably already been validated), and keeping `G1Element` objects as `bytes48` makes it clear that it's dumb deserialization that's been done and users have to ensure validation is done elsewhere.

In particular, when we start doing deserialization in rust, there's no easy way to get access to `G1Element` as either a python or C++ struct (and deserializing as a python object means grabbing the GIL which now ties the rust API to python in an unacceptable way), so keeping it around as `bytes48` is an obvious choice.

richardkiss commented on chia/consensus/block_body_validation.py:
```diff
@@ -199,8 +192,7 @@ async def validate_block_body(
         assert npc_result is not None
-        cost = calculate_cost_of_program(block.transactions_generator, npc_result, constants.COST_PER_BYTE)
-        npc_list = npc_result.npc_list
+        cost = npc_result.cost
```
Right, so a full sync is actually a pretty good test (although a bit slow).

Idea: how hard would it be to add a feature that takes the existing DB and forces a revalidation of everything? That could allow "resync tests" to run without hitting the network.

arvidn replied:
I've been thinking about such tool for a while too. it beats syncing from scratch every time we want to validate a new version. I have a tool that's similar to that, that runs all the generator programs. (in fact, it needs a reviewer :) https://github.com/Chia-Network/chia-blockchain/pull/9983 ).

However, I think we would want something that actually involves the `Blockchain` class. We could probably just do what we do in tests, and feed it all the blocks in order, from a database.

richardkiss reviewed (commented):
This is too much for me to digest, but the overall idea is very strong, and all the changes look exactly like the changes you would need to make to implement the idea. So as long as we have tests that cover all the conditions, especially boundary conditions (ie. N passes but N+1 fails or vice versa), then I think we can have high confidence this is good.

arvidn commented:
I had to rebase because of a conflict.

When I moved the conditions into rust, I also added test covering them all, here (the mempool):

https://github.com/Chia-Network/chia-blockchain/blob/main/tests/core/full_node/test_mempool.py#L293

and here (in the blockchain consensus):

https://github.com/Chia-Network/chia-blockchain/blob/main/tests/blockchain/test_blockchain.py#L1723

arvidn commented:
it might make sense to land https://github.com/Chia-Network/chia-blockchain/pull/10888 first, and rerun CI for this PR.

mariano54 commented on chia/consensus/block_body_validation.py:
```diff
@@ -407,16 +403,10 @@ async def validate_block_body(
-                if fee < 0:
-                    return Err.RESERVE_FEE_CONDITION_FAILED, None
-                assert_fee_sum = uint128(assert_fee_sum + fee)
+    assert_fee_sum: uint64 = uint64(0)
```
I think this needs to be uint128 since the max coin amount is uint64 but we might have multiple coins. not that anyone will pay a fee of < 18 million XCH though :P

arvidn replied:
There's a consensus rule saying the sum of all `RESERVE_FEE` conditions may not exceed `uint64`. (technically, it says the sum of all fees aren't allowed to exceed uint64, but that implies that the reserve_fee also cannot exceed).

The current implementation enforcing this is here: https://github.com/Chia-Network/clvm_rs/blob/0.1.20/src/gen/conditions.rs#L435
The original implementation of this rule is here: https://github.com/Chia-Network/chia-blockchain/blob/1.0.0/src/consensus/block_body_validation.py#L355

mariano54 commented on chia/consensus/block_body_validation.py:
```diff
@@ -436,20 +426,24 @@ async def validate_block_body(
-            npc.condition_dict,
+    # verify absolute/relative height/time conditions
+    if npc_result is not None:
+        assert npc_result.conds
```
`assert npc_result.conds is not None`

I avoid `assert X` unless X is a bool, because it is highly confusion for an optional of an empty list.

mariano54 commented on chia/consensus/multiprocess_validation.py:
```diff
@@ -123,7 +123,8 @@ def batch_pre_validate_blocks(
                     if validate_signatures:
                         if npc_result is not None and block.transactions_info is not None:
-                            pairs_pks, pairs_msgs = pkm_pairs(npc_result.npc_list, constants.AGG_SIG_ME_ADDITIONAL_DATA)
+                            assert npc_result.conds
```
is not None

mariano54 reviewed (dismissed):
This seems like a significant simplification. The only comments are small nits

arvidn commented:
addressed review comments (in a separate commit) and rebased to fix conflict with `main`

arvidn committed:
address review comments
